### PR TITLE
Fix #200

### LINKED ISSUE #200: Request feature to read typed keys
The crate already supports determining if a particular key is down or not, but this is not sufficient to accept text input in a natural way. There are two specific goals that can't be fulfilled:

 - When a key is pressed and held, there will be many typing events in a particular OS-defined pattern. To demonstrate this, hold down any letter on your keyboard and notice that the time between the first and second characters that are typed is different from after the 3rd, 4th, etc. Also, note that the exact intervals are OS-defined, and are often configurable in system settings.
 - In some cases, someone may want to type a character that requires a complicated series of key presses, in an OS-defined way. For example, consider the Linux compose key. When it is set to right alt, one can press right alt, then a, then ' to type á or right alt then ? then ? again to type ¿, where ? is typed with the shift key. There are many thousands of such combinations. 

A proper event or mechanism to look at _typed_ keys, as opposed to pressed keys, is needed to capture this complicated and OS-specific behavior. To provide this, I propose a new method on the `minifb::Window` struct, `fn get_next_typed_char() -> Option<char>` that, when called repeatedly, will give all characters that have been typed in the order that they were typed, then `None` once all types characters have been returned. This method could then be polled frequently (e.g. every frame) to get new information, that is simply passed along from the underlying OS-specific API. Note, however, that this is just one way to implement this, and that another method may be more convenient given the platform-specific underlying APIs (which are unknown to me).

### PR DESCRIPTION
**Pull request by john01dav**
#200 is important to me so I started working on a fix for it (it appears to not happen in windows, so I just need to make the X11 behavior more consistent with the correct behavior on Windows), but I ran into a road block that I described [here](https://users.rust-lang.org/t/why-does-xcreateic-in-xlib-return-null-over-ffi-but-not-in-c/67322). Is there any chance that I could get some help here? Also, I know that the current code on this branch is a mess -- don't worry, I'll clean it up quite a bit once it's working.

### CODE REVIEW

File: src/os/posix/x11.rs [hunk 1]
```diff
@@ -384,6 +388,25 @@ impl Window {
                 &mut attributes,
             );
 
+            let empty_string = CString::new("").unwrap();
```
Comment (emoon) [inline src/os/posix/x11.rs:391]: this can be replace with `let empty_string = b"\0";` instead

File: src/os/posix/x11.rs [hunk 2]
```diff
@@ -1024,8 +1049,25 @@ impl Window {
...
-            callback.add_char(code_point);
+            callback.add_char(char);
+        }
+    }
+
+    unsafe fn code_point_to_char(&mut self, event: &mut XKeyEvent, code_point: u32) -> u32{
+        const BUFFER_SIZE: usize = 32;
+
+        unsafe {
+            let mut buff: Box<[u8]> = vec![0; BUFFER_SIZE].into_boxed_slice(); //TODO: extra heap allocation
```
Comment (emoon) [inline src/os/posix/x11.rs:1062]: Also here remove this heap allocation with something like ` let mut buffer: [u8; 32] = [0; 32];`

Comment (john01dav) [reply src/os/posix/x11.rs:1062]: This was on my list of things to get done, I just wanted to get it working first. Normally I would not do a PR until I have such things cleaned up, but I need help, hence the early PR.

**Comment by emoon** [comment]
Hey,

Thanks for the PR. What exactly do you need help with? I haven't done much in this area so I'm not sure of how much assistance I can be.

**Comment by john01dav** [comment]
> Hey,
> 
> Thanks for the PR. What exactly do you need help with? I haven't done much in this area so I'm not sure of how much assistance I can be.

I need help figuring out why XCreateIC is returning null and fixing it.

**Comment by emoon** [comment]
> I need help figuring out why XCreateIC is returning null and fixing it. (I can probably fix it myself once I know why it's happening)

I have never used this API so I'm not sure what could be wrong here

**Comment by john01dav** [comment]
@emoon It's all ready to merge now. With regard to the bug I asked for help on, it turned out that I just made a small error in the C FFI for XCreateIC.

File: src/os/posix/x11.rs [hunk 3]
```diff
@@ -384,6 +393,44 @@ impl Window {
                 &mut attributes,
             );
 
+            //let empty_string = CString::new("").unwrap();
```
Comment (emoon) [inline src/os/posix/x11.rs:396]: Remove this line

File: Cargo.toml [hunk 4]
```diff
@@ -49,7 +49,8 @@ wayland-cursor = {version = "0.28", optional = true}
 tempfile = {version = "3.2", optional = true}
 xkb = {version = "0.2.1", optional = true}
 xkbcommon-sys = {version = "0.7.5", optional = true}
-x11-dl = {version = "2.18.3", optional = true}
+x11-dl = {version = "2.19.1", optional = true}
+libc = {version = "0.2.107", optional = true}
```
Comment (emoon) [inline Cargo.toml:53]: What is `libc` used for? It would be good if no additional dependencies gets added.

Comment (john01dav) [reply Cargo.toml:53]: On line 120 of x11.rs I'm using it to set the current locale. This is needed for compose keys to work. I do not know why. To verify this, you can compile my branch on Linux (I'm testing on Debian 11), enable a compose key (a web search will show how to do this in whatever desktop environment you use), and then try to type a character with the compose key into the char_callback in my branch (I added a printout of what unicode character the code point corresponds to). If your compose key is right alt then right alt then a then ' types á. With the locale-setting line present, this works as intended. When the locale-setting line is commented out, no character is typed.

I don't think that it's ideal to have the locale set at a global basis like this, however I am not aware of any better solution. It's also worth noting that libc is already pulled in as a recursive/transitive dependency with the current dependency list, so adding it as a direc dependency is almost a no-op.

File: src/os/posix/x11.rs (hunk 3 above)
Comment (john01dav) [reply src/os/posix/x11.rs]: Done, sorry for missing that.

File: Cargo.toml (hunk 4 above)
Comment (emoon) [reply Cargo.toml:53]: Alright. Fair enough, the reason I'm asking is that I want to try to reduce the big dependency list at some point as stated in #237

**Comment by emoon** [comment]
Thanks!

**Pull request merged**